File: 1004 Max Consecutive Ones III/1004max-consecutive-ones-iii.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int ans = INT_MIN;
        int left = 0;
        int right = 0;
        queue<int> q1;
        int ze = 0;
        for(right;right<nums.size();right++){
            if(nums[right]==0){
                ze++;
                q1.push(right);
            }
            if(ze>k){
                left = max(left,q1.front()+1);
                q1.pop();
                ze--;
            }
            ans = max(ans,right-left+1);
        }
        return ans;
    }
};

File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:

        '''
        Pattern - Sliding Window

        TC - O(N)
        SC - 0(1)
        '''
        
        left = 0
        ans = 0
        cnt = 0

        for right in range(len(nums)):
            if nums[right] == 0:
                cnt += 1
            

            while cnt > k:
                if nums[left] == 0:
                    cnt -= 1

                left += 1
            
            ans = max(ans, right - left + 1)
    
        return ans